### PR TITLE
raftstore: reduce read snap file times

### PR DESCRIPTION
**Pull request by BusyJay**
When applying snapshot, validation will read the whole file at first, and read from the beginning again to copy the data into rocksdb. This pr combines these two steps and just reads the file once.

@siddontang @hhkbp2 PTAL

### CODE REVIEW

File: src/raftstore/store/snap.rs [hunk 1]
```diff
@@ -171,13 +148,25 @@ impl SnapFile {
         Ok(())
     }
 
+    /// Same as `raw_save`, but will automatically append the checksum to
+    /// the end of file.
+    pub fn save(&mut self) -> io::Result<()> {
```
Comment (siddontang) [inline src/raftstore/store/snap.rs]: when do we need `save` now?

File: src/raftstore/store/snap.rs [hunk 2]
```diff
@@ -221,6 +210,74 @@ impl Drop for SnapFile {
...
+    res: Option<u32>,
+}
+
+impl SnapValidationReader {
+    /// Open the snap file located at specified path.
+    fn open<P: AsRef<Path>>(path: P) -> io::Result<SnapValidationReader> {
+        let reader = try!(File::open(path.as_ref()));
+        let digest = Digest::new(crc32::IEEE);
+        let len = try!(reader.metadata()).len();
+        if len < 4 {
```
Comment (siddontang) [inline src/raftstore/store/snap.rs]: use a const var for 4

File: src/raftstore/store/snap.rs (hunk 1 above)
Comment (BusyJay) [inline src/raftstore/store/snap.rs]: When generating a snap file.

**Comment by hhkbp2** [comment]
Please merge the master.

File: src/raftstore/store/worker/region.rs [hunk 3]
```diff
@@ -253,6 +252,7 @@ impl<T: MsgSender> Runner<T> {
                     box_try!(self.db.write(wb));
                     break;
                 }
+                box_try!(util::check_key_in_region(keys::origin_key(&key), &region));
```
Comment (siddontang) [inline src/raftstore/store/worker/region.rs]: can we meet this ?

Comment (BusyJay) [inline src/raftstore/store/worker/region.rs]: If the snapfile is not correct, any key is possible to be decoded from the file.

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @hhkbp2

File: src/raftstore/store/snap.rs [hunk 4]
```diff
@@ -221,6 +212,75 @@ impl Drop for SnapFile {
...
+            reader: reader,
+            digest: digest,
+            left: left,
+            res: None,
+        })
+    }
+
+    /// Validate the file
+    ///
+    /// If the reader will be comsumed after calling this method, no further data can be
```
Comment (hhkbp2) [inline src/raftstore/store/snap.rs]: s/comsumed/consumed/

File: src/raftstore/store/snap.rs [hunk 5]
```diff
@@ -171,13 +150,25 @@ impl SnapFile {
...
+    /// the end of file.
+    pub fn save(&mut self) -> io::Result<()> {
+        self.save_impl(true)
+    }
+
     /// Use the content in temporary files replace the target file.
     ///
     /// Please note that this method can only be called once.
-    pub fn save(&mut self) -> io::Result<()> {
+    pub fn raw_save(&mut self) -> io::Result<()> {
```
Comment (hhkbp2) [inline src/raftstore/store/snap.rs]: FYI, I prefer to use
save -> nothing more than just flush and close the file.
save_with_checksum -> a longer name compared to 'save', which means more work to do: to append checksum, flush and close file.

File: src/raftstore/store/snap.rs [hunk 6]
```diff
@@ -508,4 +568,33 @@ mod test {
...
+        f1.save().unwrap();
+        let mut reader = f1.reader().unwrap();
+        reader.validate().unwrap();
+
+        // partial read should not affect validation.
+        reader = f1.reader().unwrap();
+        reader.read(&mut [0, 0]).unwrap();
+        reader.validate().unwrap();
+
+        // fully read should not affect validation.
```
Comment (hhkbp2) [inline src/raftstore/store/snap.rs]: 'fully' is an 'adverb'.

**Comment by hhkbp2** [comment]
Rest LGTM

**Comment by siddontang** [comment]
Ping @BusyJay

**Comment by BusyJay** [comment]
PTAL

**Comment by hhkbp2** [comment]
LGTM

**Pull request merged**